### PR TITLE
refactor(modules): ES5 require statements converted to ES2015 import statements

### PR DESCRIPTION
**Pull request by farhan787**
issue #532

### CODE REVIEW

**Comment by farhan787** [comment]
@shane-tomlinson could you please close #691 PR.

**Comment by farhan787** [comment]
@shane-tomlinson we can't make changes in following directories/files,
1 . scripts/lib/ because it belongs to the backend side and things aren't working with `import`.

2 . few files in app/tests/ read [this](https://github.com/mozilla/fxa/blob/e09cd808d36f586dd0021ac58a1c371b048068a5/packages/fxa-content-server/app/tests/test_start.js#L11) and some other files are giving errors and functional tests are not running because of them because they're using backend stuff.

3 . scripts/vendor because it isn't suppose to change now.

Also, I know that 2 functional tests are failing inside [this file](https://github.com/mozilla/fxa/blob/e09cd808d36f586dd0021ac58a1c371b048068a5/packages/fxa-content-server/app/tests/spec/views/sms_sent.js#L83), may be because I changed these [2 lines](https://github.com/mozilla/fxa/blob/e09cd808d36f586dd0021ac58a1c371b048068a5/packages/fxa-content-server/app/scripts/views/sms_sent.js#L18-L19). Do you have any opinion on how to handle these two lines ? You can look the changes I did in my commit for these lines whether they are correct or need modification.

File: packages/fxa-content-server/app/scripts/views/mixins/upgrade-session-mixin.js [hunk 1]
```diff
@@ -18,16 +18,16 @@
 
 'use strict';
 
-const BaseView = require('../base');
+import BaseView from '../base';
 const { preventDefaultThen } = BaseView;
```
[the remark this answers is not in the document]
Comment (farhan787) [reply packages/fxa-content-server/app/scripts/views/mixins/upgrade-session-mixin.js]: Sure

**Comment by farhan787** [comment]
@shane-tomlinson I hope you like the changes :-)

**Comment by farhan787** [comment]
> Screenshot 2019-04-10 at 07 00 40

@shane-tomlinson I actually made the changes in all of these files and believe me the functional tests were failing and after debugging I had to change the `import` to `require` in these files and only then functional tests were passing.

**Comment by shane-tomlinson** [comment]
@farhan787 I merged a huge PR earlier that removed several files, you'll need to rebase this off of master and remove the same files.

**Comment by shane-tomlinson** [comment]
> @shane-tomlinson I actually made the changes in all of these files and believe me the functional tests were failing and after debugging I had to change the import to require in these files and only then functional tests were passing.

The important change that I don't see is the webpack.config.js change mentioned above, without it  the necessary transforms won't happen on the unit tests.

The other important change is to replace the [spread operators in the unit tests](https://github.com/mozilla/fxa/blob/6b71bdbf6d2af7ccfbf5e1138e737254edaefc83/packages/fxa-content-server/app/tests/spec/lib/experiments/grouping-rules/token-code.js#L121) with [Object.assign](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign). 

I then made these transformations in this order:

* `const assert = require('chai').assert;` => `import { assert } from 'chai'`

Then 
* `const (.*) = require\(([^)]*)\)` => `import $1 from $2`

**Comment by farhan787** [comment]
> The other important change is to replace the [spread operators in the unit tests](https://github.com/mozilla/fxa/blob/6b71bdbf6d2af7ccfbf5e1138e737254edaefc83/packages/fxa-content-server/app/tests/spec/lib/experiments/grouping-rules/token-code.js#L121) with [Object.assign](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign).

@shane-tomlinson could you please elaborate this one, I'm trying to do it like this,
```js
Object.assign({}, subject),                 // and I think this isn't right way to do it
```

**Comment by shane-tomlinson** [comment]
> @shane-tomlinson could you please elaborate this one, I'm trying to do it like this,

Here's an example:

```diff
diff --git a/packages/fxa-content-server/app/tests/spec/lib/experiments/grouping-rules/index.js b/packages/fxa-content-server/app/tests/spec/lib/experiments/grouping-rules/index.js
index d9b9974ba..5f873fc35 100644
--- a/packages/fxa-content-server/app/tests/spec/lib/experiments/grouping-rules/index.js
+++ b/packages/fxa-content-server/app/tests/spec/lib/experiments/grouping-rules/index.js
@@ -66,13 +66,12 @@ describe('lib/experiments/grouping-rules/index', () => {
 
       assert.isTrue(experimentGroupingRules.choose('rule1', subject));
       assert.isTrue(rule1.choose.calledOnce);
-      assert.deepEqual(rule1.choose.args[0][0], {
+      assert.deepEqual(rule1.choose.args[0][0], Object.assign({
         experimentGroupingRules,
         featureFlags: {
           foo: 'bar'
         },
-        ...subject
-      });
+      }, subject));
 
       assert.equal(experimentGroupingRules.choose('rule2', subject), 'treatment');
       assert.isTrue(rule2.choose.calledOnce);
```

**Comment by farhan787** [comment]
> Here's an example:

Thanks for the example @shane-tomlinson

**Comment by farhan787** [comment]
> I then made these transformations in this order:
> 
> * `const assert = require('chai').assert;` => `import { assert } from 'chai'`
> 
> Then
> 
> * `const (.*) = require\(([^)]*)\)` => `import $1 from $2`

@shane-tomlinson though this is done, but this regex is invalid for this. It's returning this,

`import assert from 'chai'.assert`

File: packages/fxa-content-server/app/tests/spec/lib/experiments/grouping-rules/token-code.js [hunk 2]
```diff
@@ -79,8 +79,7 @@ describe('lib/experiments/grouping-rules/token-code', () => {
 
       it('featureFlags take precedence', () => {
         subject.clientId = 'invalidClientId';
-        assert.equal(experiment.choose({
-          ...subject,
+        assert.equal(experiment.choose(Object.assign({
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/tests/spec/lib/experiments/grouping-rules/token-code.js]: Perfect!

File: packages/fxa-content-server/app/tests/spec/views/settings/clients.js [hunk 3]
```diff
@@ -4,20 +4,20 @@
 
 'use strict';
 
-const $ = require('jquery');
+import $ from 'jquery';
 const _ = require ('underscore');
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/tests/spec/views/settings/clients.js]: We missed one here! Let's fix that as a follow on.

File: packages/fxa-content-server/webpack.config.js [hunk 4]
```diff
@@ -111,7 +111,8 @@ const webpackConfig = {
       {
         test: /\.js$/,
         include: [
-          path.resolve(__dirname, 'app', 'scripts')
+          path.resolve(__dirname, 'app', 'scripts'),
+          path.resolve(__dirname, 'app', 'tests')
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/webpack.config.js]: 👍

File: packages/fxa-content-server/app/tests/spec/views/settings/clients.js (hunk 3 above)
Comment (farhan787) [reply packages/fxa-content-server/app/tests/spec/views/settings/clients.js]: @shane-tomlinson I don't know why grep missed it https://pastebin.com/vceLcK54

Comment (farhan787) [reply packages/fxa-content-server/app/tests/spec/views/settings/clients.js]: but I've converted it too.

File: packages/fxa-content-server/app/tests/spec/lib/experiments/grouping-rules/token-code.js (hunk 2 above)
Comment (farhan787) [reply packages/fxa-content-server/app/tests/spec/lib/experiments/grouping-rules/token-code.js]: Thanks.

**Comment by shane-tomlinson** [comment]
@farhan787 can you rebase this once more? #829 merged, so the tests should all pass now.

**Comment by shane-tomlinson** [comment]
Reopened as #838, rebased from the current master. Closing this one.

**Comment by shane-tomlinson** [comment]
Re-opening this and closing mine, I'd rather @farhan787 rebase and he can have the glory of seeing this merge.

**Comment by shane-tomlinson** [comment]
Re-opened as #843 so that I could sign the commit.

**Pull request closed without merging**